Add policy to bump postgres minimum version (#7709)
mattermost/docs: source/install/prepare-mattermost-database.rst

.. important::

  If you are upgrading a major version of Postgres, it is essential that ``ANALYZE VERBOSE`` is run on the database post upgrade. This is necessary to re-populate the ``pg_statistics`` table used to generate optimal query plans. The database performance might suffer if this step is not done.

Minimum supported version policy
---------------------------------

To make planning easier and ensure your Mattermost deployment remains fast and secure, we are introducing a policy for updating the minimum supported version of Postgres. The oldest supported Postgres version Mattermost supports will match the oldest version supported by the Postgres community. This ensures you benefit from the latest features and security updates.

This policy change takes effect from Mattermost v10.6, where the minimum Postgres version required will be Postgres 13. This aligns with the Postgres community's support policy, which provides 5 years of support for each major version.

.. note::

>   Mattermost v10.6 is not an :ref:`Extended Support Release (ESR) <about/release-policy:extended support releases>`. Going forward, this database version support policy will only apply to ESR releases. When a Postgres version reaches its end of life (EOL), Mattermost will require a newer version starting with the next scheduled ESR release.

Reviewer: Wondering if having a bullet list or table with a map of the future support plan could be helpful (to us and customers alike). For example, if v10.11 is the next v10 ESR, we plan to bump Postgres min version to v14.

Reviewer 2: I like this suggestion.

Reviewer 3: @amyblais - Which server versions between v10.5 and v11 are ESRs? I'll incorporate these details into this docs PR.

Reviewer 3: Confirmed with Eric that v10.5 and v.11 be our next ESRs ahead of v11. Given that this policy is being introduced with v10.6, the next ESR release would be v10.11, and that's where we'd bump the minimum version of PostgreSQL to v14. I'll incorporate this update into this docs PR.

> Customers will have 9 months to plan, test, and upgrade their Postgres version before the new requirement takes effect. If you need more time, extended ESR support may be available on a case-by-case basis. This policy aims to provide clarity and transparency so you can align database upgrades with the Mattermost release schedule. Contact a `Mattermost Expert <https://mattermost.com/contact-sales/>`_. to discuss your options.

Reviewer 2: I think we should remove the "If you need more time..." sentence.  While this *may* be true, I don't think we want to publicize it.

> Customers will have 9 months to plan, test, and upgrade their Postgres version before the new requirement takes effect. If you need more time, extended ESR support may be available on a case-by-case basis. This policy aims to provide clarity and transparency so you can align database upgrades with the Mattermost release schedule. Contact a `Mattermost Expert <https://mattermost.com/contact-sales/>`_. to discuss your options.

Reviewer 3: ```suggestion
Customers will have 9 months to plan, test, and upgrade their PostgreSQL version before the new requirement takes effect. This policy aims to provide clarity and transparency so you can align database upgrades with the Mattermost release schedule. Contact a `Mattermost Expert <https://mattermost.com/contact-sales/>`_. to discuss your options.
```

  If you are upgrading a major version of Postgres, it is essential that ``ANALYZE VERBOSE`` is run on the database post upgrade. This is necessary to re-populate the ``pg_statistics`` table used to generate optimal query plans. The database performance might suffer if this step is not done.

Minimum supported version policy
---------------------------------

To make planning easier and ensure your Mattermost deployment remains fast and secure, we are introducing a policy for updating the minimum supported version of PostgreSQL. The oldest supported PostgreSQL version Mattermost supports will match the oldest version supported by the PostgreSQL community. This ensures you benefit from the latest features and security updates.

This policy change takes effect from Mattermost v10.6, where the minimum PostgreSQL version required will be PostgreSQL 13. This aligns with the PostgreSQL community's support policy, which provides 5 years of support for each major version.

.. note::

  - Mattermost v10.6 is not an :ref:`Extended Support Release (ESR) <about/release-policy:extended support releases>`. Going forward, this database version support policy will only apply to ESR releases.

>   - When a PostgreSQL version reaches its end of life (EOL), Mattermost will require a newer version starting with the next scheduled ESR release. This means the following future PostgreSQL minimum version increases as follows:

Reviewer 3: @sadohert - Is this what you had in mind?

Reviewer 2: I think this works.  Combining these 2 resources:

https://www.postgresql.org/support/versioning/

https://docs.mattermost.com/about/mattermost-server-releases.html

I was imagining a table like:

Mattermost ESR | Release Date | Minimum Postgres Version
-- | -- | --
9.11 | Aug 16,2025 | 12.x
10.5 | Feb 16,2025 | 13.x
10.11 | Aug 16, 2025 | 13.x
11.5 * | Feb 16, 2026 | 14.x

\* Forecasted version and date

Author: @sadohert - One concern for the table is that it can lead to further maintenance effort for the team. Because now it's one more item to keep track of and update every ESR. I was thinking that just the 2 lines would be sufficient for customers to get an idea and extrapolate. WDYT?

Reviewer 3: @agnivade - I share the same concerns about maintenance. @sadohert - Are you open to publishing what we have drafted and iterate if we start receiving feedback asking for further clarity?

Reviewer: I feel the maintenance effort required to update a single row in the table is justified in this case as it adds valuable clarity for customers and eliminates any potential ambiguity. We could consider incorporating this as part of a playbook's checklist for the ESR release process."

Reviewer 3: @streamer45 - Understood. Please confirm the next 3-4 Mattermost server ESR versions and the associated PG minimum supported version increases please? I'll introduce the table to the docs. I ask for the confirmation because I _think_ we want v10.6 (not an ESR)/PG 13.x, v10.11/PG 14.x, and v11.5/PG 15x, but I could be wrong. Looking to nail this down, then update the docs. Thanks!

Reviewer: @cwarnermm Working on top of what @sadohert provided, I think this would be accurate:

Mattermost Version | Release Date | Minimum Postgres Version
-- | -- | --
9.11 ESR | Aug 2025 | 11.x
10.5 ESR | Feb 2025 | 11.x
10.6 | March 2025 | 13.x
10.11 ESR | Aug 2025 | 13.x
11.5 ESR * | Feb 2026 | 14.x

\* Forecasted version and date

@agnivade Please confirm as well.

Reviewer 2: Agreed on maintenance issue... my hope was we could do this:

> consider incorporating this as part of a playbook's checklist for the ESR release process.

Once a year it'll be a `nop` action, and once a year it'll be an update to the table.

Author: On 11.5 ESR, we will still be using 13.x. Because 14.x goes EOL on 12 Nov 2026.

Reviewer 2: My bad.  I was using the "Final Release" date on the Postgres docs as the EOL date... 

How do we know the actual EOL?

Author: https://endoflife.date/postgresql

Reviewer 3: @sadohert - I agree with a release playbook maintenance step to ensure this guidance remains accurate. @amyblais - How do you recommend we approach this, given that it isn't something we need to review on a monthly basis?

Reviewer 2: I'm still seeing the minimum PG version for `11.5`  as Postgres `14`.

Postgres 13 goes EOL on Nov. 13th/2025, according to https://endoflife.date/postgresql

So this coming November PG 13 goes EOL... I thought our policy was that the next ESR will bump... and thus our minimum goes to 14?

@agnivade ?

Author: Oops, my brain stopped functioning there. You are right @sadohert.

Reviewer 4: I'll add this step to the release playbook.

Reviewer 3: Thanks, @amyblais!

@sadohert - I believe this PR is ready for your final review.

   - Mattermost will require PostgreSQL 14 from the Mattermost v10.11 ESR scheduled for August 2025.

>    - Mattermost will require PostgreSQL 15 from the Mattermost v11.5 ESR scheduled for February 2026.

Reviewer 3: ```suggestion
   - Mattermost will require PostgreSQL 15 from the Mattermost v11.5 ESR forecasted for February 2026.
```

+-----------------------------------------------------------+------------------+--------------------------------+
| **Mattermost Version**                                    | **Release Date** | **Minimum PostgreSQL Version** |
+===========================================================+==================+================================+
| :ref:`v9.11 ESR <release-v9-11-extended-support-release>` | 2024-8-15        | 11.x                           |
+-----------------------------------------------------------+------------------+--------------------------------+
| v10.5 ESR                                                 | 2025-2-15        | 11.x                           |
+-----------------------------------------------------------+------------------+--------------------------------+
| v10.6                                                     | 2025-3-15        | 13.x                           |
+-----------------------------------------------------------+------------------+--------------------------------+

> | v10.11 ESR                                                | 2025-7-15        | 13.x                           |

Reviewer 2: I think we are forecasting an ESR in August 2025 (6 months after the Feb 2025 ESR).

So the date for v10.11 should be 2025-08-15

Reviewer 2: @amyblais  can you confirm that forecasting?  Things can change, but if we follow the pattern from the last year this would be the timeline.

Reviewer 4: v10.11 ESR on 2025-08-15 is correct.

> | v10.11 ESR                                                | 2025-7-15        | 13.x                           |

Reviewer 3: ```suggestion
| v10.11 ESR                                                | 2025-8-15        | 13.x                           |
```

> | v10.11 ESR                                                | 2025-8-15        | 13.x                           |

Reviewer 3: ```suggestion
| v10.11 ESR                                                | 2025-8-15        | 14.x                           |
```

> | v10.11 ESR                                                | 2025-8-15        | 14.x                           |

Reviewer 3: ```suggestion
| v10.11 ESR                                                | 2025-8-15        | 13.x                           |
```

+-----------------------------------------------------------+------------------+--------------------------------+

> | v11.5 ESR ``*``                                           | 2026-2-15        | 13.x (EOL 12-11-2026)          |

Reviewer 3: ```suggestion
| v11.5 ESR ``*``                                           | 2026-2-15        | 14.x (EOL 2025-11-13)          |
```

> | v11.5 ESR ``*``                                           | 2026-2-15        | 14.x (EOL 2025-11-13)          |

Reviewer: PSQL v14 goes EOL November 12, 2026

> | v11.5 ESR ``*``                                           | 2026-2-15        | 14.x (EOL 2025-11-13)          |

Reviewer 3: ```suggestion
| v11.5 ESR ``*``                                           | 2026-2-15        | 14.x (EOL 2026-11-12)          |
```
